### PR TITLE
misc: fix memory leak under per-vci

### LINKED ISSUE #5200: bug/threads: memory leak for multi-threaded test under multi-vci
Some of the affected tests:
`mt_bsendrecv, mt_ssendrecv, mt_isendirecv, mt_ibsendirecv, mt_issendirecv, mt_mprobe_sendrecv, mt_probe_sendrecv_huge`

Output:
```
Unexpected output in mt_isendirecv: [1] 2048 at [0x7f87cc001bc0], er_mpich/mpich-ofi/src/include/mpir_handlemem.h[167]
Unexpected output in mt_ibsendirecv: [1] 491520 at [0x7f24195420b0], er_mpich/mpich-ofi/src/include/mpir_handlemem.h[186]
```

Corresponding code locations:
```
In MPIR_Handle_indirect_init(..):
            objmem->indirect =
                (void **) MPL_calloc(indirect_num_blocks, sizeof(void *), MPL_MEM_OBJECT);
...
            block_ptr = (void *) MPL_calloc(indirect_num_indices, objmem->size, MPL_MEM_OBJECT);

```

Configuration: num of vci > 1, no-odd-even, debug build

Helpful info to replicate the issue:
* Config string: 
```
'--with-device=ch4:ofi' '--disable-ft-tests' '--with-libfabric=embedded' '--disable-mlx' '--enable-nemesis-dbg-localoddeven' '--enable-thread-cs=per-vci' '--with-ch4-max-vcis=4' '--enable-large-tests' '--enable-collalgo-tests' '--disable-perftest' '--with-wrapper-dl-type=rpath' 'CC=gcc' 'CXX=g++' 'FC=gfortran' 'F77=gfortran' '-enable-g=all' '-enable-error-checking=all' '-enable-fast=none' '-enable-debuginfo' '-enable-ch4-netmod-inline=no' '-enable-ch4-shm-inline=no' 'MPICHLIB_CFLAGS=-O0 -Wall -ggdb  -ggdb -Wall -mtune=generic -std=gnu99' 'MPICHLIB_CXXFLAGS=-O0 -Wall -ggdb  -ggdb -Wall -mtune=generic' 'MPICHLIB_FCFLAGS=-O0 -Wall -ggdb  -ggdb -Wall -mtune=generic' 'MPICHLIB_F77FLAGS=-O0 -Wall -ggdb  -ggdb -Wall -mtune=generic' 'MPICHLIB_LDFLAGS=-O0 -mtune=generic' 'CC=gcc' 'LDFLAGS=-Wl,-z,now' 'CXX=g++' 'FC=gfortran' 'F77=gfortran'
```
* Set runtime env variable:
```
export MPIR_CVAR_ODD_EVEN_CLIQUES=1
export MPIR_CVAR_CH4_MT_MODEL=direct
export MPIR_CVAR_CH4_NUM_VCIS=4
```
* Run multi-threaded tests after building MPICH
```
cd test/mpi/threads/pt2pt/
make testing
```

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
* ch4/ofi: fix memory leak in addr_exchange_all_vnis
* init: add lock in MPIR_Add_finalize 

Fixes #5200 

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all
netmod: ch4:ofi
config: shmem

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all --enable-thread-cs=per-vci --ch4-max-vcis=4
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all --enable-thread-cs=per-vci --with-ch4-max-vcis=4
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all --enable-thread-cs=per-vci --with-ch4-max-vcis=4
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all --enable-thread-cs=per-vci --with-ch4-max-vcis=4
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --enable-g=all --enable-thread-cs=per-vci --with-ch4-max-vcis=4
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

File: src/mpi/init/init_util.c [hunk 1]
```diff
@@ -21,12 +21,15 @@ typedef struct Finalize_func_t {
...
 #define MAX_FINALIZE_FUNC 64
+static MPL_initlock_t fstack_lock = MPL_INITLOCK_INITIALIZER;
 static Finalize_func_t fstack[MAX_FINALIZE_FUNC];
 static int fstack_sp = 0;
 static int fstack_max_priority = 0;
 
 void MPIR_Add_finalize(int (*f) (void *), void *extra_data, int priority)
 {
+    /* MPIR_Add_finalize may be called both inside and outside init */
+    MPL_initlock_lock(&fstack_lock);
```
Comment (sagarth) [inline src/mpi/init/init_util.c:32]: This should behave as a global lock, right? Could it come in the critical path and impact the intended concurrency for the per-vci mode?

Comment (sagarth) [reply src/mpi/init/init_util.c:32]: Okay. Then this PR looks good to me.

**Review by sagarth — APPROVED** [review]

**Review by sagarth — APPROVED** [review]

**Review by raffenet — APPROVED** [review]

**Pull request merged**